oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -183,87 +195,182 @@ pub struct Runtime {
     aux_servers: Mutex<Vec<AuxServer>>,
 }
 
+// Methods which translate between ABI handles (Node-relative u64 values) and `ChannelHalf`
+// values.
 impl Runtime {
-    /// Creates a [`Runtime`] instance but does not start executing any Node.
-    pub fn create(configuration: Configuration) -> Self {
-        Self {
+    fn new_abi_handle(&self, node_id: NodeId, half: ChannelHalf) -> oak_abi::Handle {
+        let mut node_infos = self.node_infos.write().unwrap();
+        let node_info = node_infos.get_mut(&node_id).expect("Invalid node_id");
+        loop {
+            let candidate = rand::thread_rng().next_u64();
+            if node_info.abi_handles.get(&candidate).is_none() {
+                debug!(
+                    "{:?}: new ABI handle {} maps to {:?}",
+                    node_id, candidate, half
+                );
+                node_info.abi_handles.insert(candidate, half);
+                return candidate;
+            }
+        }
+    }
+    // Remove the handle from the Node's handle table.
+    fn drop_abi_handle(&self, node_id: NodeId, handle: oak_abi::Handle) -> Result<(), OakStatus> {
+        let mut node_infos = self.node_infos.write().unwrap();
+        let node_info = node_infos.get_mut(&node_id).expect("Invalid node_id");
+        node_info
+            .abi_handles
+            .remove(&handle)
+            .ok_or(OakStatus::ErrBadHandle)
+            .map(|_half| ())
+    }
+    fn abi_to_half(
+        &self,
+        node_id: NodeId,
+        handle: oak_abi::Handle,
+    ) -> Result<ChannelHalf, OakStatus> {
+        let node_infos = self.node_infos.read().unwrap();
+        let node_info = node_infos.get(&node_id).expect("Invalid node_id");
+        let half = node_info
+            .abi_handles
+            .get(&handle)
+            .ok_or(OakStatus::ErrBadHandle)?;
+        trace!("{:?}: map ABI handle {} to {:?}", node_id, handle, half);
+        Ok(half.clone())
+    }
+    fn abi_to_read_half(
+        &self,
+        node_id: NodeId,
+        handle: oak_abi::Handle,
+    ) -> Result<ChannelHalf, OakStatus> {
+        let half = self.abi_to_half(node_id, handle)?;
+        match half.direction {
+            ChannelHalfDirection::Read => Ok(half),
+            ChannelHalfDirection::Write => Err(OakStatus::ErrBadHandle),
+        }
```

Comment:
This feels like a (fallible) downcast operation (even more so with the enum representation for channel that I suggested elsewhere), perhaps worth trying to encapsulate that in its own method at some point.

Reply:
For me, this method *is* the encapsulation of the fallible operation.  The user has given us an ABI handle for a read operation, but they may have messed up and either passed in:
 - a number that's not a handle
 - a number that is a handle, but isn't for the read side of a channel.

Once this method is passed, calling code has a reference to a channel that is guaranteed to be valid and readable.

(I'm not sure that the shift to an enum would make this much nicer either, at least without one more level of indirection in the type, because it's not possible to return something of type `Reference::ReadChannelHalf`.)